pkg/reconciler/taskrun/taskrun_test.go

```diff
@@ -350,6 +353,47 @@ func eventFromChannel(c chan string, testName string, wantEvents []string) error
 	return nil
 }
 
+// eventFromChannelUnordered takes a chan of string and a list of events that a test
+// expects to receive. The events can be received in any order. Any extra or too few
+// events are both considered errors.
+func eventFromChannelUnordered(c chan string, wantEvents []string) error {
+	timer := time.NewTimer(1 * time.Second)
+	expected := append([]string{}, wantEvents...)
+	// loop len(expected) + 1 times to catch extra erroneous events received that the test is not expecting
+	maxEvents := len(expected) + 1
+	for eventCount := 0; eventCount < maxEvents; eventCount++ {
+		select {
+		case event := <-c:
+			if len(expected) == 0 {
+				return fmt.Errorf("extra event received: %q", event)
+			}
+			found := false
+			for wantIdx, want := range expected {
+				matching, err := regexp.MatchString(want, event)
+				if err != nil {
+					return fmt.Errorf("something went wrong matching an event: %s", err)
+				}
+				if matching {
+					found = true
+					// Remove event from list of those we expect to receive
+					expected[wantIdx] = expected[len(expected)-1]
+					expected = expected[:len(expected)-1]
+					break
+				}
+			}
+			if !found {
+				return fmt.Errorf("unexpected event received: %q", event)
+			}
+		case <-timer.C:
+			if len(expected) != 0 {
+				return fmt.Errorf("timed out waiting for %d more events", len(expected))
```

Comment:
maybe include the remaining expected events too

fmt.Errorf("timed out waiting for %d more events: %#v", len(expected), expected)

Reply:
Good idea; done!